### PR TITLE
migrate to go modules

### PR DESCRIPTION
**Pull request by vrothberg**
Note that the CNI_COMMIT in the Makefile now evaluates to something
looking like that: v0.0.0-20190506202554-9467ac9cfd92.  The reason
for the format change is because the `go.mod` file is the only source
of truth and go simply stores certain version information like that.

Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

### CODE REVIEW

**Comment by vrothberg** [comment]
The time has come :smiling_imp: Once this PR is in, we can use dependabot for opening PRs to update dependencies.

**Comment by TomSweeneyRedHat** [comment]
Looks like a dependency is messed up somewhere.

```
go: finding github.com/mvdan/interfacer latest
can't load package: package github.com/mvdan/interfacer/cmd/interfacer: unknown import path "github.com/mvdan/interfacer/cmd/interfacer": cannot find module providing package github.com/mvdan/interfacer/cmd/interfacer
WARNING: failed to install github.com/mvdan/interfacer/cmd/interfacer: exit status 1
```

**Comment by vrothberg** [comment]
There's some more work needed before we can use go modules (mostly CI-related). I'd love to get https://github.com/golangci/golangci-lint into the CI as we already did for CRI-O. The metalinter one seems to be abandoned.

**Comment by rhatdan** [comment]
@vrothberg Any movement on this?

**Comment by vrothberg** [comment]
> @vrothberg Any movement on this?

Nope. We need `golangci-lint` in before.

File: Makefile [hunk 1]
```diff
@@ -74,10 +74,10 @@ endif
...
 .PHONY: install.tools
 install.tools:
-	$(GO) get -u $(BUILDFLAGS) github.com/cpuguy83/go-md2man
-	$(GO) get -u $(BUILDFLAGS) github.com/vbatts/git-validation
-	$(GO) get -u $(BUILDFLAGS) github.com/onsi/ginkgo/ginkgo
-	$(GO) get -u $(BUILDFLAGS) gopkg.in/alecthomas/gometalinter.v1
+	GO111MODULE=off $(GO) get -u $(BUILDFLAGS) github.com/cpuguy83/go-md2man
+	GO111MODULE=off $(GO) get -u $(BUILDFLAGS) github.com/vbatts/git-validation
+	GO111MODULE=off $(GO) get -u $(BUILDFLAGS) github.com/onsi/ginkgo/ginkgo
+	GO111MODULE=off $(GO) get -u $(BUILDFLAGS) gopkg.in/alecthomas/gometalinter.v1
```
Comment (saschagrunert) [inline Makefile]: I would vendor the tools within the repo to have reproducible outputs from them.

Comment (vrothberg) [reply Makefile]: Doing that will make it _very_ hard for downstream packagers to figure out which dependency is needed for Buildah and which for it's CI. Debian, for instance, needs to package **all** dependencies in separate .deb packages. I think @siretart will shoot at me :^) Maybe, we can workaround that by creating another module.

Comment (vrothberg) [reply Makefile]: Yes, that should work. Maybe we can use the same trick for CRI-O to move the CI deps out of the main `go.mod`?

Comment (saschagrunert) [reply Makefile]: I was wondering how this influences Debian packaging. If every (go) dependency is part of the repo then it's not a dependency anymore from the perspective of packaging, right? :grimacing: 

Yes, the idea would be to add a `tools` package, which will never build at all: https://github.com/cri-o/cri-o/blob/master/tools/tools.go 
The go modules discovery via `go list` will add them to `go.mod`.

Comment (vrothberg) [reply Makefile]: Debian can't use the `./vendor` folder, and all dependencies have to be packaged separately in to a .deb. Mixing build and test dependencies into one module would complicate update a lot as the downstream packagers would have to figure out which is which.

**Comment by vrothberg** [comment]
@TomSweeneyRedHat @rhatdan @edsantiago @saschagrunert PTAL

**Comment by vrothberg** [comment]
Splitting the build and test dependencies into separate modules could also be done for libpod (and cri-o).

File: Makefile [hunk 2]
```diff
@@ -62,23 +62,19 @@ deps: gopath
...
+validate: install.tools
 	# Run gofmt on version 1.11 and higher
 ifneq ($(GO110),$(GOVERSION))
 	@./tests/validate/gofmt.sh
 endif
 	@./tests/validate/whitespace.sh
 	@./tests/validate/govet.sh
 	@./tests/validate/git-validation.sh
-	@./tests/validate/gometalinter.sh . cmd/buildah
+	./tests/tools/build/golangci-lint run
```
Comment (TomSweeneyRedHat) [inline Makefile]: I'd rather add golangci-lint as a separate PR.

Comment (vrothberg) [reply Makefile]: Sure, can do that.

**Comment by vrothberg** [comment]
@TomSweeneyRedHat, CI is finally green and we're back on gometalinter. The PR can trivially be rebased now once we're ready for the migration.

File: go.sum [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+github.com/docker/docker v0.7.3-0.20180827131323-0c5f8d2b9b23 h1:mJtkfC9RUrUWHMk0cFDNhVoc9U3k2FRAzEZ+5pqSIHo=
+github.com/docker/docker v0.7.3-0.20180827131323-0c5f8d2b9b23/go.mod h1:eEKB0N0r5NX/I1kEveEz05bcu8tLC/8azJZsviup8Sk=
+github.com/docker/docker-credential-helpers v0.6.1 h1:Dq4iIfcM7cNtddhLVWe9h4QDjsi4OER3Z8voPu/I52g=
+github.com/docker/docker-credential-helpers v0.6.1/go.mod h1:WRaJzqw3CTB9bk10avuGsjVBZsD05qeibJ1/TYlvc0Y=
+github.com/docker/go-connections v0.4.0 h1:El9xVISelRB7BuFusrZozjnkIM5YnzCViNKohAFqRJQ=
+github.com/docker/go-connections v0.4.0/go.mod h1:Gbd7IOopHjR8Iph03tsViu4nIes5XhDvyHbTtUxmeec=
+github.com/docker/go-units v0.3.2/go.mod h1:fgPhTUdO+D/Jk86RDLlptpiXQzgHJF7gydDDbaIK4Dk=
+github.com/docker/go-units v0.3.3 h1:Xk8S3Xj5sLGlG5g67hJmYMmUgXv5N4PhkjJHHqrwnTk=
+github.com/docker/go-units v0.3.3/go.mod h1:fgPhTUdO+D/Jk86RDLlptpiXQzgHJF7gydDDbaIK4Dk=
+github.com/docker/libnetwork v0.0.0-20181207044626-1a06131fb8a0/go.mod h1:93m0aTqz6z+g32wla4l4WxTrdtvBRmVzYRkYvasA5Z8=
```
Comment (edsantiago) [inline go.sum]: Likewise

**Comment by vrothberg** [comment]
> About ten more, but I'll hold off on entering those unless they're actually a problem (I'm guessing they're not).

The source of truth is the `go.mod` file. The `go.sum` one is containing hashes of currently and previously used versions/revisions, so we can safely ignore it.

I just deleted `go.sum` and re-ran `make vendor` which resulted in the following diff:
```diff
--- a/go.sum
+++ b/go.sum
@@ -26,17 +26,14 @@ github.com/cyphar/filepath-securejoin v0.2.1/go.mod h1:FpkQEhXnPnOthhzymB7CGsFk2
 github.com/davecgh/go-spew v1.1.1/go.mod h1:J7Y8YcW2NihsgmVo/mv3lAwl/skON4iLHjSsI+c5H38=
 github.com/docker/distribution v0.0.0-20170817175659-5f6282db7d65 h1:4zlOyrJUbYnrvlzChJ+jP2J3i77Jbhm336NEuCv7kZo=
 github.com/docker/distribution v0.0.0-20170817175659-5f6282db7d65/go.mod h1:J2gT2udsDAN96Uj4KfcMRqY0/ypR+oyYUYmja8H+y+w=
-github.com/docker/docker v0.0.0-20190307005417-54dddadc7d5d/go.mod h1:eEKB0N0r5NX/I1kEveEz05bcu8tLC/8azJZsviup8Sk=
 github.com/docker/docker v0.7.3-0.20180827131323-0c5f8d2b9b23 h1:mJtkfC9RUrUWHMk0cFDNhVoc9U3k2FRAzEZ+5pqSIHo=
 github.com/docker/docker v0.7.3-0.20180827131323-0c5f8d2b9b23/go.mod h1:eEKB0N0r5NX/I1kEveEz05bcu8tLC/8azJZsviup8Sk=
 github.com/docker/docker-credential-helpers v0.6.1 h1:Dq4iIfcM7cNtddhLVWe9h4QDjsi4OER3Z8voPu/I52g=
 github.com/docker/docker-credential-helpers v0.6.1/go.mod h1:WRaJzqw3CTB9bk10avuGsjVBZsD05qeibJ1/TYlvc0Y=
 github.com/docker/go-connections v0.4.0 h1:El9xVISelRB7BuFusrZozjnkIM5YnzCViNKohAFqRJQ=
 github.com/docker/go-connections v0.4.0/go.mod h1:Gbd7IOopHjR8Iph03tsViu4nIes5XhDvyHbTtUxmeec=
-github.com/docker/go-units v0.3.2/go.mod h1:fgPhTUdO+D/Jk86RDLlptpiXQzgHJF7gydDDbaIK4Dk=
 github.com/docker/go-units v0.3.3 h1:Xk8S3Xj5sLGlG5g67hJmYMmUgXv5N4PhkjJHHqrwnTk=
 github.com/docker/go-units v0.3.3/go.mod h1:fgPhTUdO+D/Jk86RDLlptpiXQzgHJF7gydDDbaIK4Dk=
-github.com/docker/libnetwork v0.0.0-20181207044626-1a06131fb8a0/go.mod h1:93m0aTqz6z+g32wla4l4WxTrdtvBRmVzYRkYvasA5Z8=
 github.com/docker/libnetwork v0.8.0-dev.2.0.20180608203834-19279f049241 h1:+ebE/hCU02srkeIg8Vp/vlUp182JapYWtXzV+bCeR2I=
 github.com/docker/libnetwork v0.8.0-dev.2.0.20180608203834-19279f049241/go.mod h1:93m0aTqz6z+g32wla4l4WxTrdtvBRmVzYRkYvasA5Z8=
 github.com/docker/libtrust v0.0.0-20160708172513-aabc10ec26b7 h1:UhxFibDNY/bfvqU5CAUmr9zpesgbU6SWc8/B4mflAE4=
@@ -91,10 +88,8 @@ github.com/onsi/ginkgo v1.6.0/go.mod h1:lLunBs/Ym6LB5Z9jYTR76FiuTmxDTDusOGeTQH+W
 github.com/onsi/gomega v1.4.1/go.mod h1:C1qb7wdrVGGVU+Z6iS04AVkA3Q65CEZX59MT0QO5uiA=
 github.com/onsi/gomega v1.4.3 h1:RE1xgDvH7imwFD45h+u2SgIfERHlS2yNG4DObb5BSKU=
 github.com/onsi/gomega v1.4.3/go.mod h1:ex+gbHU/CVuBBDIJjb2X0qEXbFg53c61hWP/1CpauHY=
-github.com/opencontainers/go-digest v0.0.0-20180430190053-c9281466c8b2/go.mod h1:cMLVZDEM3+U2I4VmLI6N8jQYUd2OVphdqWwCJHrFt2s=
 github.com/opencontainers/go-digest v1.0.0-rc1 h1:WzifXhOVOEOuFYOJAW6aQqW0TooG2iki3E3Ii+WN7gQ=
 github.com/opencontainers/go-digest v1.0.0-rc1/go.mod h1:cMLVZDEM3+U2I4VmLI6N8jQYUd2OVphdqWwCJHrFt2s=
-github.com/opencontainers/image-spec v1.0.0/go.mod h1:BtxoFyWECRxE4U/7sNtV5W15zMzWCbyJoFRP3s7yZA0=
 github.com/opencontainers/image-spec v1.0.1 h1:JMemWkRwHx4Zj+fVxWoMCFm/8sYGGrUVojFA6h/TRcI=
 github.com/opencontainers/image-spec v1.0.1/go.mod h1:BtxoFyWECRxE4U/7sNtV5W15zMzWCbyJoFRP3s7yZA0=
 github.com/opencontainers/runc v0.1.1/go.mod h1:qT5XzbpPznkRYVz/mWwUaVBUv2rmF59PVA73FjuZG0U=
@@ -159,13 +154,10 @@ golang.org/x/sync v0.0.0-20181221193216-37e7f081c4d4 h1:YUO/7uOKsKeq9UokNS62b8FY
 golang.org/x/sync v0.0.0-20181221193216-37e7f081c4d4/go.mod h1:RxMgew5VJxzue5/jJTE5uejpjVlOe/izrB70Jof72aM=
 golang.org/x/sys v0.0.0-20180824143301-4910a1d54f87/go.mod h1:STP8DvDyc/dI5b8T5hshtkjS+E42TnysNCUPdjciGhY=
 golang.org/x/sys v0.0.0-20180909124046-d0be0721c37e/go.mod h1:STP8DvDyc/dI5b8T5hshtkjS+E42TnysNCUPdjciGhY=
-golang.org/x/sys v0.0.0-20190108104531-7fbe1cd0fcc2/go.mod h1:STP8DvDyc/dI5b8T5hshtkjS+E42TnysNCUPdjciGhY=
 golang.org/x/sys v0.0.0-20190422165155-953cdadca894 h1:Cz4ceDQGXuKRnVBDTS23GTn/pU5OE2C0WrNTOYK1Uuc=
 golang.org/x/sys v0.0.0-20190422165155-953cdadca894/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
-golang.org/x/text v0.0.0-20181227161524-e6919f6577db/go.mod h1:bEr9sfX3Q8Zfm5fL9x+3itogRgK3+ptLWKqgva+5dAk=
 golang.org/x/text v0.3.0 h1:g61tztE5qeGQ89tm6NTjjM9VPIm088od1l6aSorWRWg=
 golang.org/x/text v0.3.0/go.mod h1:NqM8EUOU14njkJ3fqMW+pc6Ldnwhi/IjpwHt7yyuwOQ=
-golang.org/x/tools v0.0.0-20180917221912-90fa682c2a6e/go.mod h1:n7NCudcB/nEzxVGmLbDWY5pfWTLqBcC2KZ6jyYvM4mQ=
 gopkg.in/airbrake/gobrake.v2 v2.0.9/go.mod h1:/h5ZAUhDkGaJfjzjKLSjv6zCL6O0LLBxU4K+aSYdM/U=
 gopkg.in/check.v1 v0.0.0-20161208181325-20d25e280405/go.mod h1:Co6ibVJAznAaIkqp8huTwlJQCZ016jof/cbN4VW5Yz0=
 gopkg.in/fsnotify.v1 v1.4.7 h1:xOHLXZwVvI9hhs+cLKq5+I5onOuwQLhQwiu63xxlHs4=
```

**Comment by vrothberg** [comment]
Thanks a lot for checking btw, @edsantiago :pray:

**Comment by edsantiago** [comment]
@vrothberg thanks for your explanation. Comparing vendor.conf and go.mod, and ignoring the many newly-added packages and handful of removed ones, the only diff that stands out to me is:
```
github.com/opencontainers/selinux: vendor=v1.1 gomod=v0.0.0-20190118194635-b707dfcb00a1
```
If I understand go modules correctly, `v1` is an important indicator of stability. Can I assume that your choice to use a `v0` here is intentional? And would it merit a git commit message (or comment in go.mod)?

This is very minor and should not in any block you from merging.

**Comment by vrothberg** [comment]
> @vrothberg thanks for your explanation. Comparing vendor.conf and go.mod, and ignoring the many newly-added packages and handful of removed ones, the only diff that stands out to me is:
> 
> ```
> github.com/opencontainers/selinux: vendor=v1.1 gomod=v0.0.0-20190118194635-b707dfcb00a1
> ```
> 
> If I understand go modules correctly, `v1` is an important indicator of stability. Can I assume that your choice to use a `v0` here is intentional? And would it merit a git commit message (or comment in go.mod)?

Awesome catch, @edsantiago! I fixed it by moving to selinux v1.2.2. Not sure how it ended up like that.

**Comment by edsantiago** [comment]
Thank you. The rest of the changes look OK to me, but I'll include them below for your review anyway:
```
Removed from vendor.conf:
  github.com/Azure/go-ansiterm
  github.com/Microsoft/go-winio
  github.com/Nvveen/Gotty
  github.com/docker/docker
  github.com/docker/go-connections
  github.com/gorilla/context
  github.com/gorilla/mux
  github.com/hashicorp/errwrap
  golang.org/x/text
Added to go.mod:
  github.com/davecgh/go-spew
  github.com/fsnotify/fsnotify
  github.com/golang/protobuf
  github.com/google/go-cmp
  github.com/hpcloud/tail
  github.com/inconshreveable/mousetrap
  github.com/konsorten/go-windows-terminal-sequences
  github.com/onsi/ginkgo
  github.com/pmezard/go-difflib
  github.com/stretchr/objx
  github.com/stretchr/testify
  github.com/vishvananda/netlink
  github.com/vishvananda/netns
  gopkg.in/airbrake/gobrake.v2
  gopkg.in/check.v1
  gopkg.in/fsnotify.v1
  gopkg.in/gemnasium/logrus-airbrake-hook.v2
  gopkg.in/tomb.v1
  gotest.tools
Diff:
  github.com/docker/go-units                           v0.3.2 -> v0.3.3
  github.com/docker/libnetwork           1a06131fb8a047d919f7 -> v0.8.0-dev.2.0.20180608203834-19279f049241
  github.com/opencontainers/go-digest    c9281466c8b2f606084a -> v1.0.0-rc1
  github.com/opencontainers/image-spec                 v1.0.0 -> v1.0.1
  github.com/opencontainers/selinux                      v1.1 -> v1.2.2
  github.com/seccomp/containers-golang                   v0.1 -> v0.0.0-20180629143253-cdfdaa7543f4
  golang.org/x/sys                       7fbe1cd0fcc20051e1fc -> v0.0.0-20190422165155-953cdadca894
  k8s.io/client-go                       kubernetes-1.10.13-b -> v0.0.0-20181219152756-3dd551c0f083
```

**Comment by vrothberg** [comment]
I restarted the one flaked Travis job.

> Thank you. The rest of the changes look OK to me, but I'll include them below for your review anyway:

I think we need to compare the `./vendor` tree as well. `go-ansiterm`, for instance, is not mentioned in `go.mod` but in `go.sum` and it's in `./vendor`. To be honest, I don't yet know why...

**Comment by edsantiago** [comment]
Chained dependency? It's listed in `vendor/github.com/fsouza/go-dockerclient/go.mod`. IIRC that's one of the rationales behind go modules, the ability to bring in chained dependencies without API breaks.

**Comment by vrothberg** [comment]
@edsantiago, it looks like it's one from Windows :^)

**Comment by edsantiago** [comment]
Tests are green. LGTM. This was a doozy, thanks for your hard work on it.

**Comment by rhatdan** [comment]
Need WIP label removed.

**Comment by vrothberg** [comment]
> Need WIP label removed.

Done :heavy_check_mark: I will prepare follow-up PRs to update the dependencies once this one's merged.

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**